### PR TITLE
http-instrumentation: fix propagation errors when using Expect header #481

### PR DESCRIPTION
**Pull request by vmarchaud**

### CODE REVIEW

File: packages/opencensus-instrumentation-http/src/http.ts [hunk 1]
```diff
@@ -360,12 +360,28 @@ export class HttpPlugin extends BasePlugin {
 
       const setter: HeaderSetter = {
         setHeader(name: string, value: string) {
-          request.setHeader(name, value);
+          // https://github.com/googleapis/cloud-trace-nodejs/pull/766
```
Comment (draffensperger) [inline packages/opencensus-instrumentation-http/src/http.ts]: What would you think about removing this PR reference and instead putting a short explanation of the logic?

File: packages/opencensus-instrumentation-http/src/http.ts [hunk 2]
```diff
@@ -360,12 +360,28 @@ export class HttpPlugin extends BasePlugin {
...
 
       const propagation = plugin.tracer.propagation;
       if (propagation) {
+        // If outgoing request headers contain the "Expect" header, the returned
+        // ClientRequest will throw an error if any new headers are added. For
+        // this reason, only in this scenario, we opt to clone the options
+        // object to inject the trace context header instead of using
+        // ClientRequest#setHeader. (We don't do this generally because cloning
+        // the options object is an expensive operation.)
+        // https://github.com/googleapis/cloud-trace-nodejs/pull/766
```
Comment (draffensperger) [inline packages/opencensus-instrumentation-http/src/http.ts]: The explanation above is great - can we remove the PR reference? (People can use the blame layer to find it if needed)

**Comment by vmarchaud** [comment]
@draffensperger I've rewritten the explanation and removed the link to the google cloud PR

**Comment by draffensperger** [comment]
Looks good, thanks!

File: packages/opencensus-instrumentation-http/test/test-http.ts [hunk 3]
```diff
@@ -374,6 +375,25 @@ describe('HttpPlugin', () => {
...
+         const options = Object.assign(
+             {headers: {Expect: '100-continue'}},
+             url.parse('http://google.fr/'));
+         await httpRequest.get(options).then((result) => {
+           assert.strictEqual(rootSpanVerifier.endedRootSpans.length, 1);
+           assert.ok(
+               rootSpanVerifier.endedRootSpans[0].name.indexOf('GET /') >= 0);
+
+           const span = rootSpanVerifier.endedRootSpans[0];
+           assertSpanAttributes(span, 301, 'GET', 'google.fr', '/', undefined);
```
Comment (mayurkale22) [inline packages/opencensus-instrumentation-http/test/test-http.ts]: Can you remove this `undefined`, the `userAgent` is an optional param.

File: packages/opencensus-instrumentation-http/src/http.ts [hunk 4]
```diff
@@ -360,12 +360,28 @@ export class HttpPlugin extends BasePlugin {
...
         }
       };
 
       const propagation = plugin.tracer.propagation;
       if (propagation) {
+        // If outgoing request headers contain the "Expect" header, the returned
+        // ClientRequest will throw an error if any new headers are added.
+        // So we need to clone the options object to be able to inject new
+        // header.
+        if (options.headers && plugin.hasExpectHeader(options)) {
```
Comment (mayurkale22) [inline packages/opencensus-instrumentation-http/src/http.ts]: Is it safe to remove `options.headers` check as we are doing same in `hasExpectHeader` function?

**Comment by vmarchaud** [comment]
Done @mayurkale22

**Comment by mayurkale22** [comment]
Thanks for the contribution @vmarchaud

**Pull request merged**